codegen/tests/type_generic_param.rs

```diff
@@ -0,0 +1,39 @@
+use delegation::delegate;
+
+#[delegate]
+trait Named<N> {
+    fn name(&self) -> N;
+}
+
+struct User(String);
+impl Named<String> for User {
+    fn name(&self) -> String {
+        self.0.clone()
+    }
+}
+
+#[delegate(derive(Named<String>))]
+enum Case1<U>
+where
+    U: Named<String> + 'static,
+{
+    User(U),
+}
+
+#[delegate(derive(
+    for<U> Named<String>
+    where
+        U: Named<String> + 'static;
+))]
+enum Case2<U> {
```

Comment:
@50U10FCA7 `for<U>` is redundant here too. Why repeat it it if it's defined on the enum itself already?

Reply:
@tyranron Because we want to check `where` closure expansion.

Reply:
@50U10FCA7 I don't see any problem with that check. The parameter is still in scope.

Reply:
@tyranron This simplification assumes that generic parameters of the type (or trait in case of `delegate(for(..))`), the macro expands will be used on resulting
```rust
impl<AllGens> Trait<TraitGens> for Type<TypeGens> {}
```

And in case `for<..>` is provided it replaces `impl<AllGens>`.

So to check whether `where` clause expands we need to specify `for<..>` in `#[delegate(derive(..))]` arguments, which requires us to place `U` generic to make:
```rust
impl<U> Named<String> for Case2<U>
where
    U: Named<String>
{}
```
impl.

Reply:
@50U10FCA7 no, we don't need to make user specifying obvious things. We could easily deduct what should be in the expansion for the parameters of the type. Concretely, this example doesn't suffer in any way from the following syntax:
```rust
#[delegate(derive(
    Named<String> where U: Named<String> + 'static;
))]
enum Case2<U> {
    Admin(U),
}
```

Reply:
@tyranron 

> Concretely, this example doesn't suffer in any way from the following syntax:
> 
> ```rust
> #[delegate(derive(
>     Named<String> where U: Named<String> + 'static;
> ))]
> enum Case2<U> {
>     Admin(U),
> }
> ```
>
> can you provide the example which suffers fro such change?

We can omit `for<..>` here without suffering anything. The only reason `for<..>` should be specified to place `where` here is to make the expression follow HKG syntax (`for<..> where`).

Reply:
@50U10FCA7 

> The only reason `for<..>` should be specified to place `where` here is to make the expression follow HKG syntax (`for<..> where`).

I don't see why user should follow this syntax.